src/OpenTelemetry.Instrumentation.EventCounters/EventCounterListener.cs

```diff
@@ -0,0 +1,229 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Diagnostics.Tracing;
+using System.Reflection;
+
+namespace OpenTelemetry.Instrumentation.EventCounters
+{
+    internal class EventCounterListener : EventListener
+    {
+        internal static readonly AssemblyName AssemblyName = typeof(EventCounterListener).Assembly.GetName();
+        internal static readonly string InstrumentationName = AssemblyName.Name;
+        internal static readonly string InstrumentationVersion = AssemblyName.Version.ToString();
+
+        private readonly bool isInitialized = false;
+        private readonly Meter meter;
+        private readonly EventCounterMetricsOptions options;
+        private readonly ConcurrentDictionary<MetricKey, Instrument> metricInstruments = new();
+        private readonly ConcurrentDictionary<MetricKey, long> lastLongValue = new();
+        private readonly ConcurrentDictionary<MetricKey, double> lastDoubleValue = new();
+        private readonly ConcurrentBag<EventSource> eventSources = new();
+
+        public EventCounterListener(EventCounterMetricsOptions options)
+        {
+            this.meter = new Meter(InstrumentationName, InstrumentationVersion);
+            this.options = options ?? throw new ArgumentNullException(nameof(options));
+
+            this.isInitialized = true;
+            this.EnablePendingEventSources(); // Some OnEventSourceCreated may have fired before constructor, enable them
+        }
+
+        private enum InstrumentType
+        {
+            Gauge,
+            Counter,
+        }
+
+        private Dictionary<string, string> EnableEventArgs => new Dictionary<string, string> { ["EventCounterIntervalSec"] = this.options.RefreshIntervalSecs.ToString(), };
+
+        protected override void OnEventSourceCreated(EventSource source)
+        {
+            // TODO: Add Configuration options to selectively subscribe to EventCounters
+            try
+            {
+                if (!this.isInitialized)
+                {
+                    this.eventSources.Add(source);
+                }
+                else
+                {
+                    this.EnableEvents(source, EventLevel.Verbose, EventKeywords.All, this.EnableEventArgs);
+                }
+            }
+            catch (Exception ex)
+            {
+                EventCountersInstrumentationEventSource.Log.ErrorEventCounter(source.Name, ex.Message);
+            }
+        }
+
+        protected override void OnEventWritten(EventWrittenEventArgs eventData)
+        {
+            if (!this.isInitialized || !eventData.EventName.Equals("EventCounters"))
```

Comment:
Where does the "EventCounters" name come from?

Reply:
It is hardhoded in the implementation class:  https://github.com/dotnet/runtime/blob/e026f392b2bb93bc9509ac5403d3e8cd7d3eee72/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventCounter.cs#L129